File: sample/output/vector.cpp
#include "vector.h"
vec2 subtract(const vec2& lhs, const vec2& rhs);
float squared_magnitude(const vec2& v);
float squared_distance(const vec2& lhs, const vec2& rhs) {
    vec2 diff = subtract(lhs, rhs);
    return squared_magnitude(diff);
}
vec2 subtract(const vec2& lhs, const vec2& rhs) {
    vec2 result;
    (result.x = (lhs.x - rhs.x));
    (result.y = (lhs.y - rhs.y));
    return result;
}
float squared_magnitude(const vec2& v) {
    return ((v.x * v.x) + (v.y * v.y));
}

File: sample/vector.ml
module vector;

export vec2: struct = {
    x: f32;
    y: f32;
}

export squared_distance: func = (lhs: vec2, rhs: vec2) -> f32 {
    diff: vec2 = subtract(lhs, rhs);
    return squared_magnitude(diff);
}

subtract: func = (lhs: vec2, rhs: vec2) -> vec2 {
    result: vec2;
    result.x = lhs.x - rhs.x;
    result.y = lhs.y - rhs.y;

    return result;
}

squared_magnitude: func = (v: vec2) -> f32 {
    return v.x * v.x + v.y * v.y;
}